### PR TITLE
 feat: implement relaxed period in clique

### PR DESCRIPTION
**Pull request by omerfirmak**
## 1. Purpose or design rationale of this PR

Block times are limiting our throughput because we can only put so many txns in a block due to circuit limits. The short term solution is to allow more frequent blocks.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Comment by omerfirmak** [comment]
depends on #735, see the second commit

File: consensus/clique/clique.go [hunk 1]
```diff
@@ -328,7 +328,7 @@ func (c *Clique) verifyCascadingFields(chain consensus.ChainHeaderReader, header
 	if parent == nil || parent.Number.Uint64() != number-1 || parent.Hash() != header.ParentHash {
 		return consensus.ErrUnknownAncestor
 	}
-	if parent.Time+c.config.Period > header.Time {
+	if parent.Time < header.Time {
```
Comment (Thegaram) [inline consensus/clique/clique.go]:
```suggestion
	if parent.Time > header.Time {
```

Comment (omerfirmak) [reply consensus/clique/clique.go]: LOL, yeah I just realized that. Fixed.

**Review by Thegaram — DISMISSED** [review]

**Review by Thegaram — DISMISSED** [review]

File: consensus/clique/clique.go [hunk 2]
```diff
@@ -328,7 +328,7 @@ func (c *Clique) verifyCascadingFields(chain consensus.ChainHeaderReader, header
 	if parent == nil || parent.Number.Uint64() != number-1 || parent.Hash() != header.ParentHash {
 		return consensus.ErrUnknownAncestor
 	}
-	if parent.Time+c.config.Period > header.Time {
+	if header.Time < parent.Time {
```
Comment (colinlyguo) [inline consensus/clique/clique.go:331]:
```suggestion
	if header.Time <= parent.Time {
```

Comment (colinlyguo) [reply consensus/clique/clique.go:331]: or is it OK to allow the same timestamp (looks viable)?

File: miner/scroll_worker.go [hunk 3]
```diff
@@ -708,7 +708,14 @@ func (w *worker) startNewPipeline(timestamp int64) {
 
 	w.currentPipelineStart = time.Now()
 	w.currentPipeline = pipeline.NewPipeline(w.chain, w.chain.GetVMConfig(), parentState, header, nextL1MsgIndex, w.getCCC()).WithBeforeTxHook(w.beforeTxHook)
-	if err := w.currentPipeline.Start(time.Unix(int64(header.Time), 0)); err != nil {
+
+	deadline := time.Unix(int64(header.Time), 0)
+	if w.chainConfig.Clique != nil && w.chainConfig.Clique.RelaxedPeriod {
+		// clique with relaxed period uses time.Now() as the header.Time, calculate the deadline
+		deadline = time.Unix(int64(header.Time+w.chainConfig.Clique.Period), 0)
```
Comment (colinlyguo) [inline miner/scroll_worker.go:715]: why set a deadline here?

File: consensus/clique/clique.go (hunk 2 above)
Comment (omerfirmak) [reply consensus/clique/clique.go:331]: Yes, it is. We might end up having sub-second block times.

File: miner/scroll_worker.go (hunk 3 above)
Comment (omerfirmak) [reply miner/scroll_worker.go:715]: With `RelaxedPeriod=false`, `header.Time` is a timestamp that is in the future and we use that as the deadline to seal the block.
But when `RelaxedPeriod=true`, `header.Time` is the time that we start building a block. So it is always in the past. That is why miner has to calculate its own deadline for a given block.

File: consensus/clique/clique.go (hunk 2 above)
Comment (colinlyguo) [reply consensus/clique/clique.go:331]: it makes sense.

File: miner/scroll_worker.go (hunk 3 above)
Comment (colinlyguo) [reply miner/scroll_worker.go:715]: got it. it makes sense. btw, I suggest that we add a comment in `clique.go` about `when RelaxedPeriod=true, header.Time is the time that we start building a block`

**Review by colinlyguo — DISMISSED** [review]

File: miner/scroll_worker.go (hunk 3 above)
Comment (omerfirmak) [reply miner/scroll_worker.go:715]: done

**Review by colinlyguo — APPROVED** [review]

**Review by Thegaram — APPROVED** [review]

**Pull request merged**